promql/functions.go

```diff
@@ -827,353 +821,103 @@ func dateWrapper(vals []Value, enh *EvalNodeHelper, f func(time.Time) float64) V
-		Call:       funcMinute,
-	},
-	"month": {
-		Name:       "month",
-		ArgTypes:   []ValueType{ValueTypeVector},
-		Variadic:   1,
-		ReturnType: ValueTypeVector,
-		Call:       funcMonth,
-	},
-	"predict_linear": {
-		Name:       "predict_linear",
-		ArgTypes:   []ValueType{ValueTypeMatrix, ValueTypeScalar},
-		ReturnType: ValueTypeVector,
-		Call:       funcPredictLinear,
-	},
-	"quantile_over_time": {
-		Name:       "quantile_over_time",
-		ArgTypes:   []ValueType{ValueTypeScalar, ValueTypeMatrix},
-		ReturnType: ValueTypeVector,
-		Call:       funcQuantileOverTime,
-	},
-	"rate": {
-		Name:       "rate",
-		ArgTypes:   []ValueType{ValueTypeMatrix},
-		ReturnType: ValueTypeVector,
-		Call:       funcRate,
-	},
-	"resets": {
-		Name:       "resets",
-		ArgTypes:   []ValueType{ValueTypeMatrix},
-		ReturnType: ValueTypeVector,
-		Call:       funcResets,
-	},
-	"round": {
-		Name:       "round",
-		ArgTypes:   []ValueType{ValueTypeVector, ValueTypeScalar},
-		Variadic:   1,
-		ReturnType: ValueTypeVector,
-		Call:       funcRound,
-	},
-	"scalar": {
-		Name:       "scalar",
-		ArgTypes:   []ValueType{ValueTypeVector},
-		ReturnType: ValueTypeScalar,
-		Call:       funcScalar,
-	},
-	"sort": {
-		Name:       "sort",
-		ArgTypes:   []ValueType{ValueTypeVector},
-		ReturnType: ValueTypeVector,
-		Call:       funcSort,
-	},
-	"sort_desc": {
-		Name:       "sort_desc",
-		ArgTypes:   []ValueType{ValueTypeVector},
-		ReturnType: ValueTypeVector,
-		Call:       funcSortDesc,
-	},
-	"sqrt": {
-		Name:       "sqrt",
-		ArgTypes:   []ValueType{ValueTypeVector},
-		ReturnType: ValueTypeVector,
-		Call:       funcSqrt,
-	},
-	"stddev_over_time": {
-		Name:       "stddev_over_time",
-		ArgTypes:   []ValueType{ValueTypeMatrix},
-		ReturnType: ValueTypeVector,
-		Call:       funcStddevOverTime,
-	},
-	"stdvar_over_time": {
-		Name:       "stdvar_over_time",
-		ArgTypes:   []ValueType{ValueTypeMatrix},
-		ReturnType: ValueTypeVector,
-		Call:       funcStdvarOverTime,
-	},
-	"sum_over_time": {
-		Name:       "sum_over_time",
-		ArgTypes:   []ValueType{ValueTypeMatrix},
-		ReturnType: ValueTypeVector,
-		Call:       funcSumOverTime,
-	},
-	"time": {
-		Name:       "time",
-		ArgTypes:   []ValueType{},
-		ReturnType: ValueTypeScalar,
-		Call:       funcTime,
-	},
-	"timestamp": {
-		Name:       "timestamp",
-		ArgTypes:   []ValueType{ValueTypeVector},
-		ReturnType: ValueTypeVector,
-		Call:       funcTimestamp,
-	},
-	"vector": {
-		Name:       "vector",
-		ArgTypes:   []ValueType{ValueTypeScalar},
-		ReturnType: ValueTypeVector,
-		Call:       funcVector,
-	},
-	"year": {
-		Name:       "year",
-		ArgTypes:   []ValueType{ValueTypeVector},
-		Variadic:   1,
-		ReturnType: ValueTypeVector,
-		Call:       funcYear,
-	},
-}
-
-// getFunction returns a predefined Function object for the given name.
-func getFunction(name string) (*Function, bool) {
-	function, ok := Functions[name]
-	return function, ok
+var FunctionCalls = map[string]FunctionCall{
```

Comment:
Having the list of functions in two places feels wrong.

Reply:
The reason for that is that the implementation of the function calls shouldn't belong to the parser package, but the type information should.

Do you have an idea for another way to achieve this?

Reply:
Not offhand, but at the least we should have a test that the two have the same elements.

Reply:
I've added such tests.